### PR TITLE
feat(cdk: popper): add cdk popper

### PR DESCRIPTION
**Pull request by LaamGinghong**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [our guidelines](https://github.com/IduxFE/idux/blob/main/packages/site/src/docs/Contributing.zh.md#commit)
- [x] Tests for the changes have been added/updated or not needed
- [x] Docs and demo have been added/updated or not needed

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Component style update
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Application (the showcase website) / infrastructure changes
- [ ] Other... Please describe:

### CODE REVIEW

**Review by danranVm — APPROVED** [review]

File: packages/cdk/popper/src/usePopper.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+  function hide(delay = state.hideDelay): void {
+    toggle(false, delay, state, setTimer)
+  }
+
+  function update(options: Partial<PopperOptions> = {}): () => void {
+    Object.assign(state, options)
+
+    return forceUpdate
+  }
```
Comment (danranVm) [inline packages/cdk/popper/src/usePopper.ts:66]:
```ts
  function update(options: Partial<PopperOptions> = {}, force = false): () => void {
    Object.assign(state, options)
   
   if(force ) {
   forceUpdate()
}
  }
```

File: packages/cdk/popper/docs/Index.zh.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+### `usePopper`
+
+```typescript
+import type { ComponentPublicInstance, ComputedRef, Ref } from 'vue'
+
+export type PopperScrollStrategy = 'close' | 'reposition' | 'none'
+export type PopperTrigger = 'click' | 'hover' | 'focus' | 'contextmenu' | 'manual'
+export type PopperElement = ComponentPublicInstance | HTMLElement
+export type PopperPlacement =
```
[the remark this answers is not in the document]
Comment (LaamGinghong) [reply packages/cdk/popper/docs/Index.zh.md:25]: 这个保持和其他参数一致吧，用驼峰

[the remark this answers is not in the document]
Comment (LaamGinghong) [reply packages/cdk/popper/docs/Index.zh.md:25]: 改

File: packages/cdk/popper/src/hooks.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  for (const [key, value] of Object.entries(options)) {
+    if (isUndefined(value)) {
+      continue
+    }
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    state[key] = value
+  }
+
+  return state
```
[the remark this answers is not in the document]
Comment (LaamGinghong) [reply packages/cdk/popper/src/hooks.ts:49]: 改

File: packages/cdk/popper/src/hooks.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  action: { show(): void; hide(): void },
+): ComputedRef<PopperTriggerEvents> {
+  const triggerEventsMap: Record<PopperTrigger, (keyof PopperTriggerEvents)[]> = {
+    click: ['onClick'],
+    focus: ['onFocus', 'onBlur'],
+    hover: ['onMouseenter', 'onMouseleave'],
+    contextmenu: ['onContextmenu'],
+    manual: [],
+  }
+  return computed(() => mapTriggerEvents(triggerEventsMap[state.trigger], state, visibility, timer, action))
```
[the remark this answers is not in the document]
Comment (LaamGinghong) [reply packages/cdk/popper/src/hooks.ts:110]: 如果换了trigger的话，这里要通过computed去重新执行绑定一遍

File: packages/cdk/popper/src/hooks.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  const onMouseleave = () => {
+    if (state.trigger !== 'hover') {
+      return
+    }
+    hide()
+  }
+
+  return computed(() => ({ onMouseenter, onMouseleave }))
```
[the remark this answers is not in the document]
Comment (LaamGinghong) [reply packages/cdk/popper/src/hooks.ts:132]: 同上

File: packages/cdk/popper/src/types.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
+import type { ComponentPublicInstance, ComputedRef, Ref } from 'vue'
+
+export type PopperScrollStrategy = 'close' | 'reposition' | 'none'
+export type PopperTrigger = 'click' | 'hover' | 'focus' | 'contextmenu' | 'manual'
+export type PopperElement = ComponentPublicInstance | HTMLElement
+export type PopperPlacement =
```
[the remark this answers is not in the document]
Comment (LaamGinghong) [reply packages/cdk/popper/src/types.ts:6]: 改

File: packages/cdk/popper/src/usePopper.ts (hunk 1 above)
Comment (LaamGinghong) [reply packages/cdk/popper/src/usePopper.ts:66]: 这里看注释，有解释

File: packages/cdk/popper/src/hooks.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (danranVm) [reply packages/cdk/popper/src/hooks.ts:110]: 那应该用 watch, 而不是用 computed。
computed 会收集所有响应式依赖，只要有一个依赖更新了，就会重新绑定。

File: packages/cdk/popper/src/usePopper.ts (hunk 1 above)
Comment (danranVm) [reply packages/cdk/popper/src/usePopper.ts:66]: 我明白你的意思，但是我觉得再去调用 update 返回的函数不太友好，比如：
```ts
const forceUpdate = update({visible:false});
forceUpdate()

// or ... 😢 
update({visible:false})();
```

而应该额外加一个参数，force，这样就可以
```ts
update({visible:false}, true);
```

File: packages/cdk/popper/docs/Index.zh.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (danranVm) [reply packages/cdk/popper/docs/Index.zh.md:25]: 这个看看有没有更合适的方案，idux 使用 camelCase, 只有在送给 prope.js 的时候转成 kebab-case

File: packages/cdk/popper/src/usePopper.ts (hunk 1 above)
Comment (LaamGinghong) [reply packages/cdk/popper/src/usePopper.ts:66]: 也行吧，但我其实更倾向于前面那种

**Pull request merged**